### PR TITLE
Extend pyOpenSSL stubs

### PR DESCRIPTION
**Pull request by mhils**
This PR greatly extends the pyOpenSSL stubs. Most importantly, it adds 
```python
def __getattr__(self, name: str) -> Any: ...  # incomplete
```
to the still-incomplete `Context` and `Connection` classes, which were previously missing.

### CODE REVIEW

File: stubs/pyOpenSSL/OpenSSL/SSL.pyi [hunk 1]
```diff
@@ -1,26 +1,153 @@
...
+SSL_CB_READ_ALERT: int
+SSL_CB_WRITE_ALERT: int
+SSL_CB_ACCEPT_LOOP: int
+SSL_CB_ACCEPT_EXIT: int
+SSL_CB_CONNECT_LOOP: int
+SSL_CB_CONNECT_EXIT: int
+SSL_CB_HANDSHAKE_START: int
+SSL_CB_HANDSHAKE_DONE: int
+
+NO_OVERLAPPING_PROTOCOLS = object()
```
Comment (srittau) [inline stubs/pyOpenSSL/OpenSSL/SSL.pyi:102]: This isn't really supported in stubs. There is a proposal to add sentinel values to the typing system, but for now the best solution is probably to do:

```suggestion
NO_OVERLAPPING_PROTOCOLS: object
```

File: stubs/pyOpenSSL/OpenSSL/SSL.pyi [hunk 2]
```diff
@@ -1,26 +1,153 @@
...
+class Error(Exception): ...
+class WantReadError(Error): ...
+class WantWriteError(Error): ...
+class WantX509LookupError(Error): ...
+class ZeroReturnError(Error): ...
+class SysCallError(Error): ...
+
+def SSLeay_version(type: int) -> str: ...
+
+class Session: ...
```
Comment (srittau) [inline stubs/pyOpenSSL/OpenSSL/SSL.pyi:113]: As Python 2 is still supported, this should be marked as a new-style class explicitly:

```suggestion
class Session(object): ...
```

(Same for the other classes below, but changing them in this PR is optional.)

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, looks mostly good to me, just two comments below.

**Comment by srittau** [comment]
Also, stubtest is unhappy, since `set_min_proto_version()` and `set_max_proto_version()` are not in pyOpenSSL 20. This also means that the version in `METADATA.toml` needs to be updated. We have two options: Wait with this PR until pyOpenSSL 21 is released and update `METADATA.toml` as well, or comment out the new functions for now with a note that they will be added in version 21.

**Comment by mhils** [comment]
Good points, thanks for the quick review! :smiley:

**Review by srittau — APPROVED** [review]

**Comment by mhils** [comment]
I fixed all stubtest complaints except for `error: OpenSSL.SSL.Context.__getattr__ is not present at runtime`. This part seems right to me though, at least I think I followed https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#incomplete-stubs correctly. Not sure if it's fatal now that everything else is fixed.

**Comment by srittau** [comment]
You can add the `OpenSSL.SSL.Context.__getattr__` to `stubs/pyOpenSSL/@tests/stubtest_allowlist.txt` to work around this.

**Comment by mhils** [comment]
Thanks for handling this so swiftly! ❤️

**Pull request merged**